native/libp2p_port/internal/subscriptions/subscriptions.go

```diff
@@ -1,49 +1,139 @@
+	topics        map[string]*pubsub.Topic
+	gsub          *pubsub.PubSub
 	port          *port.Port
 }
 
-func NewSubscriber(p *port.Port) Subscriber {
-	return Subscriber{subscriptions: make(map[string]chan struct{}), port: p}
-}
+func NewSubscriber(p *port.Port, h host.Host) Subscriber {
+	heartbeat := 700 * time.Millisecond
+	gsubParams := pubsub.DefaultGossipSubParams()
+	gsubParams.D = 8
+	gsubParams.Dlo = 6
+	gsubParams.HeartbeatInterval = heartbeat
+	gsubParams.FanoutTTL = 60 * time.Second
+	gsubParams.HistoryLength = 6
+	gsubParams.HistoryGossip = 3
+
+	thresholds := &pubsub.PeerScoreThresholds{
+		GossipThreshold:             -4000,
+		PublishThreshold:            -8000,
+		GraylistThreshold:           -16000,
+		AcceptPXThreshold:           100,
+		OpportunisticGraftThreshold: 5,
+	}
+	scoreParams := &pubsub.PeerScoreParams{
+		Topics:        make(map[string]*pubsub.TopicScoreParams),
+		TopicScoreCap: 32.72,
+		AppSpecificScore: func(p peer.ID) float64 {
+			return 0
+		},
+		AppSpecificWeight:           1,
+		IPColocationFactorWeight:    -35.11,
+		IPColocationFactorThreshold: 10,
+		IPColocationFactorWhitelist: nil,
+		BehaviourPenaltyWeight:      -15.92,
+		BehaviourPenaltyThreshold:   6,
+		BehaviourPenaltyDecay:       math.Pow(0.01, 1/float64(10*32)),
+		DecayInterval:               12 * time.Second,
+		DecayToZero:                 0.01,
+		RetainScore:                 100 * 32 * 12 * time.Second,
+	}
+
+	// TODO: add more options
+	options := []pubsub.Option{
+		pubsub.WithMessageSignaturePolicy(pubsub.StrictNoSign),
+		pubsub.WithNoAuthor(),
+		pubsub.WithMessageIdFn(utils.MsgID),
+		pubsub.WithPeerScore(scoreParams, thresholds),
+		pubsub.WithGossipSubParams(gsubParams),
+		pubsub.WithSeenMessagesTTL(550 * heartbeat),
+		pubsub.WithMaxMessageSize(10 * (1 << 20)), // 10 MB
+	}
+
+	gsub, err := pubsub.NewGossipSub(context.Background(), h, options...)
+	utils.PanicIfError(err)
 
-func (s *Subscriber) Subscribe(topic_name string) {
-	_, is_subscribed := s.subscriptions[topic_name]
-	if !is_subscribed {
-		s.subscriptions[topic_name] = make(chan struct{}, 1)
-		go SubscribeToTopic(topic_name, s.subscriptions[topic_name], s.port)
+	return Subscriber{
+		subscriptions: make(map[string]chan struct{}),
+		topics:        make(map[string]*pubsub.Topic),
+		gsub:          gsub,
+		port:          p,
 	}
 }
 
-func (s *Subscriber) Unsubscribe(topic_name string) {
-	_, is_subscribed := s.subscriptions[topic_name]
-	if is_subscribed {
-		s.subscriptions[topic_name] <- struct{}{}
-		delete(s.subscriptions, topic_name)
+func (s *Subscriber) Subscribe(topicName string, handler []byte) error {
+	_, isSubscribed := s.subscriptions[topicName]
+	if isSubscribed {
+		return errors.New("already subscribed")
+	}
+	topic, joined := s.topics[topicName]
+	if !joined {
+		var err error
+		topic, err = s.gsub.Join(topicName)
+		utils.PanicIfError(err)
+		s.topics[topicName] = topic
 	}
+	sub, err := topic.Subscribe()
+	utils.PanicIfError(err)
```

Comment:
According to the underlying code, this operation can fail only when:
- the topic was closed (this can't happen)
- some option passed failed (also can't happen)
- gossipbsub was closed (it's OK to panic here)

So it can't fail unless the application is shutting down. I'm not sure if we need to handle that case explicitly.

Reply:
Makes sense! Let's leave the panic there then.